From: Chad Monroe <chad@monroe.io>
Date: Wed, 22 Jan 2025 13:14:08 -0800
Subject: [PATCH] wifi: mt76: mt7915: ignore duplicate radar events 

If a radar event is triggered while mac80211 is in the middle of
a CSA from a prior event, some channel switch parameters can get
overwritten. When this happens the radio will fail to start after
moving to the new channel.

It seems like it would be better for mac80211 to protect itself
against this but I couldn't find a clean way to do it. 

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 mt7915/mac.c    |    2 ++
 mt7915/mcu.c    |   19 +++++++++++++++++--
 mt7915/mt7915.h |    1 +
 3 files changed, 20 insertions(+), 2 deletions(-)

--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -341,6 +341,8 @@ static void
 mt7915_mcu_rx_radar_detected(struct mt7915_dev *dev, struct sk_buff *skb)
 {
 	struct mt76_phy *mphy = &dev->mt76.phy;
+	struct mt7915_phy *phy;
+	struct cfg80211_chan_def *chandef;
 	struct mt7915_mcu_rdd_report *r;
 	u32 sku;
 
@@ -370,18 +372,31 @@ mt7915_mcu_rx_radar_detected(struct mt79
 	if (!mphy)
 		return;
 
+	phy = (struct mt7915_phy *)mphy->priv;
+	chandef = &phy->mt76->chandef;
+
+	if (phy->rdd_hw_value == chandef->chan->hw_value) {
+		dev_info(dev->mt76.dev,
+			"%s: radar detection for channel=%u in progress, ignore r->rdd_idx=%d\n",
+			__func__, phy->rdd_hw_value, r->rdd_idx);
+		return;
+	}
+
 	dev_info(dev->mt76.dev, "%s: RADAR DETECTED r->rdd_idx=%d\n", __func__, r->rdd_idx);
 	dev_dbg(dev->mt76.dev, "%s: long_detected=%u constant_prf_detected=%u staggered_prf_detected=%u\n",
 			__func__, r->long_detected, r->constant_prf_detected, r->staggered_prf_detected);
 	dev_dbg(dev->mt76.dev, "%s: radar_type_idx=%u periodic_pulse_num=%u long_pulse_num=%u hw_pulse_num=%u\n",
 			__func__, r->radar_type_idx, r->periodic_pulse_num, r->long_pulse_num, r->hw_pulse_num);
 
-	if (r->rdd_idx == MT_RDD_IDX_BACKGROUND)
+	if (r->rdd_idx == MT_RDD_IDX_BACKGROUND) {
 		cfg80211_background_radar_event(mphy->hw->wiphy,
 						&dev->rdd2_chandef,
 						GFP_ATOMIC);
-	else
+	} else {
+		phy->rdd_hw_value = chandef->chan->hw_value;
 		ieee80211_radar_detected(mphy->hw, NULL);
+	}
+
 	dev->hw_pattern++;
 }
 
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -2194,6 +2194,8 @@ int mt7915_dfs_init_radar_detector(struc
 	enum mt76_dfs_state dfs_state, prev_state;
 	int err, rdd_idx = mt7915_get_rdd_idx(phy, false);
 
+	phy->rdd_hw_value = 0;
+
 	prev_state = phy->mt76->dfs_state;
 	dfs_state = mt76_phy_dfs_state(phy->mt76);
 
--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -275,6 +275,8 @@ struct mt7915_phy {
 	s16 coverage_class;
 	u8 slottime;
 
+	u16 rdd_hw_value;
+
 	u32 trb_ts;
 
 	u32 rx_ampdu_ts;
